### PR TITLE
Drop PyMC3 compatibility in favor of PyMC v5

### PR DESCRIPTION
**Pull request by michaelosthege**
It turned out that the random walks were currently in a broken state, so I first upgraded them to the new `pm.RandomWalk` API from `pymc>=4.2.2`.

This included extracting the switchpoint detection to a function (for more maintainability) and refactoring it to use `pm.logcdf` instead of calling `scipy.stats.t.cdf`. The nice thing about that is, that it's agnostic to the innovation distribution of the random walk!

I added a test case to cover situations where Student-t random walks for switchpoint detection are used in combination with known switchpoints that result in a piecewise random walk model.

After these changes, the upgrade to PyMC v5 was seamless.

Finally I re-ran the example notebook.
There are two details that I'm still investigating, but I don't consider them blocking.
+ [x] ~The detected switchpoint with the example is one time point earlier than before 🤔~
+ [x] ~The `joint_logprob` emits a bunch of warnings about orphan RVs, but the model still works as expected 🤔~ This is a PyMC bug: https://github.com/pymc-devs/pymc/issues/6401

### CODE REVIEW

File: bletl/growth.py [hunk 1]
```diff
@@ -500,3 +456,69 @@ def fit_mu_t(
...
+    # first CDF values at all mu_t elements
+    cdf_evals = []
+    for rvname in sorted(theta_map.keys()):
+        if rvname not in pmodel.named_vars:
+            continue
+        # The random walk may be split in multiple segments.
+        # We can identify a segment from the RVOp type that created it.
+        rv = pmodel[rvname]
+        if rv.owner is None:
+            continue
```
Comment (lhelleckes) [inline bletl/growth.py:498]: Can you explain again what the owner is?

File: bletl/growth.py [hunk 2]
```diff
@@ -500,3 +456,69 @@ def fit_mu_t(
...
+        if isinstance(rv.owner.op, pm.RandomWalk.rv_type):
+            # Get a handle on the innovation dist
+            innov_dist = rv.owner.inputs[1]
+            # and calculate the innovations from the MAP estimate
+            innov = numpy.diff(theta_map[rvname])
+            # so we can calculate the logcdf and cdf evaluations.
+            logcdfs = pm.logcdf(innov_dist, innov).eval()
+            # Prepend a 0.5 because the init dist does not count into the diff.
+            cdf_evals += [0.5, *numpy.exp(logcdfs)]
```
Comment (lhelleckes) [inline bletl/growth.py:507]: Why a 0.5 here?

File: bletl/growth.py [hunk 3]
```diff
@@ -500,3 +456,69 @@ def fit_mu_t(
...
+    cdf_evals = numpy.array(cdf_evals)
+    if len(cdf_evals) != len(t_data) - 1:
+        raise Exception(
+            f"Failed to find all random walk segments. Found {len(cdf_evals)}, expected {len(t_data) - 1}."
+        )
+    # filter for the elements that lie outside of the [0.005, 0.995] interval
```
Comment (lhelleckes) [inline bletl/growth.py:513]: Minor detail, but these probs can also be different from your comment.

File: bletl/growth.py [hunk 4]
```diff
@@ -500,3 +456,69 @@ def fit_mu_t(
...
+    significance_mask = numpy.logical_or(
+        cdf_evals < (switchpoint_prob / 2),
+        cdf_evals > (1 - switchpoint_prob / 2),
+    )
+    # add these autodetected timepoints to the switchpoints-dict
+    # (ignore the first timepoint)
+    switchpoints = {}
+    for t_switch, is_switchpoint in zip(t_data, significance_mask[1:]):
```
Comment (lhelleckes) [inline bletl/growth.py:521]: shouldn't you also mask t_data by [1:] here?

**Review by lhelleckes — COMMENTED** [review]
Sorry, this is just a first. I still need to understand some things, e.g. which you dropped the cycle switch points. But maybe you can comment on those already in the mean time.

File: bletl/growth.py (hunk 1 above)
Comment (michaelosthege) [reply bletl/growth.py:498]: The `.owner` of a `TensorVariable` is the `Apply` node that created it.

The `Apply` node is created by calling an `Op` instance with some inputs, and it stores a handle to the `Op` under `.op`.

Not all named variables in a PyMC model have an `.owner`, most notably any `TensorConstant`s such as `pm.ConstantData` variables.

File: bletl/growth.py (hunk 2 above)
Comment (michaelosthege) [reply bletl/growth.py:507]: 0.5 is the CDF value when falling exactly onto the prior. A `0` would be below that threshold a few lines down..

File: bletl/growth.py (hunk 4 above)
Comment (michaelosthege) [reply bletl/growth.py:521]: This is a matter of how the switchpoint time is defined.
If anything it should be shifted by one more element, because in the notebook the detected switchpoint is actually one element too early?

This is why I added the test case where it finds exactly the right time value as expected.

**Comment by michaelosthege** [comment]
> I still need to understand some things, e.g. which you dropped the cycle switch points. But maybe you can comment on those already in the mean time.

Thanks, your question reminded me why I introduced it in the first place:
The user may provide a switchpoint time that is not identical to the time value of the nearest observation.
The `c_switchpoints_known` was needed because comparison by time value (as currently done with my changes) can lead to duplicates.

I will fix this.

File: bletl/growth.py (hunk 2 above)
Comment (michaelosthege) [reply bletl/growth.py:511]: I updated these comments to be more explanatory

File: bletl/growth.py (hunk 4 above)
Comment (michaelosthege) [reply bletl/growth.py]: With the new changes, the zip now iterates over two equally sized collections and no longer slices anything.

**Comment by michaelosthege** [comment]
I went through the switchpoint detection again and it is now behaving the same way as before:

Switchpoints are autodetected at the point preceeding the first segment of the new random walk:

image

I amended this to the original commit. You can click [Compare](https://github.com/JuBiotech/bletl/compare/c4309b68131c6a01ace9055f7f20f866d3e65df4..1b657040faff24eabfd8f31471c5c57d51f0e602) next to the force-push info above to see the changes I made.
Or you can check the overall diff, of course.

**Review by lhelleckes — APPROVED** [review]
Just some minor comments. You can merge tomorrow when you decided if you still want to tackle these issues or not.

File: bletl/growth.py [hunk 5]
```diff
@@ -176,13 +169,12 @@ def _make_random_walk(
     ----------
     name : str
         Name of the random walk variable.
+    init_dist
+        A random variable to use as the prior for innovations.
     mu : float, array-like
         Mean of the random walk.
-        If a vector is passed, only the first element should be nonzero,
```
Comment (lhelleckes) [inline bletl/growth.py:181]: Why did you delete those lines?

Comment (michaelosthege) [reply bletl/growth.py:181]: Previously this was just to get the first element of the random walk to have a different prior.

With the new `pm.RandomWalk` API one can have a completely independent distribution for the first element - the `init_dist`.
Here I used `pm.Normal`, but in principle one could specify a `pm.Exponential` for where a random walk initializes.
